### ISSUE TITLE
when i increase the deepth of transformer ,i get wrong result

### ISSUE THREAD

**Issue by 520jefferson** [marian-nmt/marian-dev#155]
When i use transformer to train chinese to english model, when i use the default depth it worked normal, but when i increase the deep of encode and decode to 6 as you said @emjotde in #116, but i get weird
when i decode i get wrong result . some of the sentence translate right, but some compeletly wrong.
For example:
Chinese sentence:
28 岁 厨师 被 发现 死 于 旧金山 一 家 商场
近日 刚 搬 至 旧金山 的 一 位 28 岁 厨师 本周 被 发现 死 于 当地 一 家 商场 的 楼梯 间 。
但 受害人 的 哥哥 表示 想 不 出 有 谁 会 想 要 加害 于 他 ， 并 称 “ 一切 终于 好 起来 了 。 ”
旧金山 验尸 官 办公室 表示 ， 周 三 早 上 于 西田 购物 中心 发现 的 尸体 确认 为 28 岁 旧金山 居民 Frank Galicia 。

Result:
i 'm sorry to hear that
in recent years , with the rapid development of information technology , information technology has become more and more important .
but if you do n't want to know what you 're doing , do n't worry .
" " no.

My setting as below, and i use 7 GPUs to train, every single GPU about 22912MiB,and i train  about Ep. 10 : Up. 194730 : Sen. 5997200 : Cost 76.46 : Time 17.56s : 16146.97 words/s : L.r. 8.5993e-05.
 is it something wrong ?
Setting:
../../build/marian \
--type transformer \
--model model.npz \
--devices 0 1 2 3 4 5 6 \
--seed 0 \
--dim-emb 512 \
--dim-rnn 1024 \
--train-set ch.cut.bpe.drop50 en.tok.lower.bpe.drop50  \
--vocabs ch.cut.bpe.drop50.pkl.json en.tok.lower.bpe.drop50.pkl.json \
--dim-vocabs 48212 33560 \
--disp-freq  10 \
--save-freq  2000 \
--learn-rate 0.0003 \
--lr-warmup 16000 \
--lr-decay-inv-sqrt 16000 \
--lr-report \
--sync-sgd \
--max-length 50 \
--optimizer adam \
--mini-batch 256 \
--maxi-batch 20 \
--enc-depth 6 \
--dec-depth 6 \
--label-smoothing 0.1 \
--transformer-heads 8  \
--transformer-dim-ffn 2048 \
--transformer-preprocess d \
--transformer-dropout 0.1 \
--transformer-dropout-attention 0 \
--transformer-postprocess dhn \
--transformer-postprocess-emb d \
--tempdir tmp \
--mini-batch-fit \
--workspace 19000 \
--clip-norm 5. \
--layer-normalization \
--after-batches 20000000 \
--log train.log.0103

**Comment by emjotde** [comment]
Looks fine.
`--dim-rnn 1024` and `--layer-normalization` aren't doing anything for the transformer model, layer-normalization is enabled by default for that model. 

How's the BLEU score? If that is OK there is no sense in trying to figure out why selected translations are bad.

**Comment by 520jefferson** [comment]
BLEU score is base.  using the wmt to score,then i get this
BLEU = 7.61, 42.5/16.7/9.3/5.6 (BP=0.549, ratio=0.625, hyp_len=33602, ref_len=53770)

**Comment by 520jefferson** [comment]
@emjotde  More worse.
I get normal result former corpus different than this and training with no deep mode (i didn't training with deep mode.
with this corpus i get abnormal result both deep mode above and no deep mode the training set as follows.
  ../../build/marian \
        --type transformer \
        --model models0112_wmt_nodeep/512-1024-ch_en.npz \
        --devices 1 2 3 4 \
        --seed 0 \
        --dim-emb 512 \
        --dim-rnn 1024 \
        --train-set wmt0103.ch.cut.bpe.drop50_c2e  wmt0103.en.tok.lower.bpe.drop50_e2c \
        --vocabs wmt0103.ch.cut.bpe.drop50_c2e.pkl.json wmt0103.en.tok.lower.bpe.drop50_e2c.pkl.json \
        --dim-vocabs 48214 33560 \
        --disp-freq  10 \
        --save-freq  2000 \
        --learn-rate 0.0003 \
        --lr-warmup 16000 \
        --lr-decay-inv-sqrt 16000 \
        --lr-report \
        --sync-sgd \
        --max-length 50 \
        --optimizer adam \
        --mini-batch 256 \
        --maxi-batch 20 \
        --transformer-heads 16 \
        --transformer-dim-ffn 4096 \
        --transformer-preprocess d \
        --transformer-dropout 0.2 \
        --transformer-dropout-attention 0 \
        --dropout-rnn 0.2 \
        --dropout-src 0.2 \
        --dropout-trg 0 \
        --tempdir tmp \
        --mini-batch-fit \
        --workspace 19000 \
        --clip-norm 5. \
        --layer-normalization \
        --after-batches 20000000 \
        --log train.log.0112

**Comment by emjotde** [comment]
I am not sure if you can actually expect shallow models to work at all. I would use at least a depth of 4. 

* How big is that corpus? 
* Any particular reasons why you use so much dropout?
* Are you getting good result for RNN models on this corpus?

**Comment by 520jefferson** [comment]
First, the number of the sentences  is  17365228.
Second, the dropout which i set according to experience of nematus, so i have not so much idea ,i just try.
Third, i don't  training RNN models with this corpus, may be i can try what you say on the type of amun (but my fellow trainging  using tensorflow maybe get RNN models, but  his result is better than what i train with type of transformer, i will check it tomorrow when he comes)

**Comment by emjotde** [comment]
OK. This is a large corpus, probably no reason to use that much regularization. The transformer model is quite particular and can get stuck easily. Experience from Nematus may betray you here.

Could you please try the default settings from here, first: https://github.com/marian-nmt/marian-examples/tree/master/transformer, with `--tied-embedding` instead of `--tied-embedding-all` as I guess your vocabs are not shared between languages.

**Comment by 520jefferson** [comment]
../build/marian \
    --model models0114_wmt_deep/512-1024-ch_en.npz --type transformer \
        --train-set /data/mei.tu/work/nmt/corpus/CWMT/CWMT-0112/wmt0103.ch.cut.bpe.drop50.addnewstest_c2e /data/mei.tu/work/nmt/corpus/CWMT/CWMT-0112/wmt0103.en.tok.lower.bpe.drop50.addnewstest_c2e \
    --max-length 100 \
        --vocabs /data/mei.tu/work/nmt/corpus/CWMT/CWMT-0112/wmt0103.ch.cut.bpe.drop50.addnewstest_c2e.pkl.json /data/mei.tu/work/nmt/corpus/CWMT/CWMT-0112/wmt0103.en.tok.lower.bpe.drop50.addnewstest_c2e.pkl.json \
    --vocabs model/vocab.ende.yml model/vocab.ende.yml \
    --mini-batch-fit -w 10000 --maxi-batch 1000 \
    --save-freq 5000 --disp-freq 500 \
        --quiet-translation \
    --beam-size 6 --normalize 0.6 \
    --log train.log.0114   \
    --enc-depth 6 --dec-depth 6 \
    --transformer-heads 8 \
    --transformer-postprocess-emb d \
    --transformer-postprocess dan \
    --transformer-dropout 0.1 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
    --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \
    --tied-embeddings \
    --devices $@ --sync-sgd --seed 1111 \
    --exponential-smoothing \
        --workspace 19000 \
    --after-batches 20000000

**Comment by emjotde** [comment]
looks good, `-w` and `--workspace` is the same.

**Comment by 520jefferson** [comment]
Yes, i've merged them, and training with 7 gpus.

**Comment by 520jefferson** [comment]
Sorry, it's my fault ,there is something wrong with the corpus , and i generate the new corpus , everything looks right now.
